bundles/org.eclipse.sirius.emfjson/src/main/java/org/eclipse/sirius/emfjson/utils/GsonEObjectSerializer.java

```diff
@@ -1040,22 +1043,48 @@ private JsonElement serializeEDataType(EObject eObject, EAttribute eAttribute) {
             if (value instanceof Collection) {
                 Collection<?> collection = (Collection<?>) value;
                 for (Object object : collection) {
-                    jsonArray.add(new JsonPrimitive(eFactoryInstance.convertToString(eAttribute.getEAttributeType(), object)));
+                    if (object == null) {
+                        jsonArray.add(JsonNull.INSTANCE);
+                    } else if (this.isEDataTypeSerializableInJson(eAttribute.getEAttributeType())) {
+                        jsonArray.add(new Gson().toJsonTree(object));
+                    } else {
+                        jsonArray.add(new JsonPrimitive(eFactoryInstance.convertToString(eAttribute.getEAttributeType(), object)));
+                    }
                 }
             }
             jsonElement = jsonArray;
         } else {
             String stringValue = eFactoryInstance.convertToString(eAttribute.getEAttributeType(), value);
-
-            if (stringValue == null) {
-                stringValue = ""; //$NON-NLS-1$
+            if (this.isEDataTypeSerializableInJson(eAttribute.getEAttributeType())) {
+                jsonElement = new Gson().toJsonTree(value);
+            } else {
+                if (stringValue == null) {
+                    jsonElement = JsonNull.INSTANCE;
+                } else {
+                    jsonElement = new JsonPrimitive(stringValue);
+                }
             }
-            jsonElement = new JsonPrimitive(stringValue);
         }
 
         return jsonElement;
     }
 
+    /**
+     * Test the given {@link EDataType} with the {@link JsonResource#OPTION_IS_EDATATYPE_SERIALIZABLE_IN_JSON_TESTER}
+     * option. If the option is not set, return false.
+     *
+     * @param eDataType
+     * @return true if the given {@link EDataType} should be serialized as a Json tree.
+     */
+    private boolean isEDataTypeSerializableInJson(EDataType eDataType) {
+        @SuppressWarnings("unchecked")
```

Comment:
`isEDataTypeSerializableInJson` -> `isSerializeEDataTypeInJson`

Reply:
The naming "isSerializeEDataTypeInJson" suggests that the given EDataType is serializable, and that based on this assumption, it poses the question whether the "Json" serialization target is a valid target for serialization. I find this naming incorrect. I prefer to keep "isEDataTypeSerializableInJson", which means exactly what the test does.

Reply:
Actually you're right on one point. The test should not say "can" but "should". I propose to rename it "shouldEDataTypeBeSerializedInJson" which reflect the real intent.

Reply:
I fixed the code accordingly.